File: Problems/139. Word Break.js
// https://leetcode.com/problems/word-break/
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function (s, wordDict) {
  let i = 0, j = 0;
  const checks = [0], visited = new Set();
  while (checks.length) {
    i = checks.shift();
    if (visited.has(i)) continue;
    for (let j = i + 1; j <= s.length; j++) {
      if (wordDict.includes(s.slice(i, j))) {
        if (j === s.length)
          return true;
        checks.push(j);
      }
    }
    visited.add(i);
  }
  return false;
};

File: Problems/139. Word Break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        dp = [False] * n
        for i in range(n):
            for word in wordDict:
                if (
                    i == len(word) - 1 and s[: i + 1] == word) 
                    or (i > len(word) - 1 and s[i - len(word) + 1 : i + 1] == word and dp[i - len(word)]):
                    dp[i] = True
        return dp[-1]
